index.js

```diff
@@ -124,30 +126,18 @@ class HandlebarsPlugin {
      * @return {Boolean} true, if a handlebars file or helper has been updated
      */
     dependenciesUpdated(compilation) {
-        if (compilation.fileTimestamps.has) { // Map
-            // webpack@4.x
-            const changedFiles = [];
-
-            const prevTimestamps = this.prevTimestamps || {};
-            const currentTimestamps = {};
-
-            compilation.fileTimestamps.forEach((timestamp, watchfile) => {
-                currentTimestamps[watchfile] = timestamp;
-                if ((prevTimestamps[watchfile] || this.startTime) < (currentTimestamps[watchfile] || Infinity)) {
-                    changedFiles.push(watchfile);
-                }
-            });
-
-            this.prevTimestamps = currentTimestamps;
-            return changedFiles.length === 0 || this.containsOwnDependency(changedFiles);
-        }
-
-        // webpack@3.x
-        const changedFiles = Object.keys(compilation.fileTimestamps).filter((watchfile) =>
-            (this.prevTimestamps[watchfile] || this.startTime) < (compilation.fileTimestamps[watchfile] || Infinity)
-        );
+        // NOTE: fileTimestamps will be an `object` or `Map` depending on the webpack version
+        const fileTimestamps = compilation.fileTimestamps;
+        const fileNames = fileTimestamps.has ? Array.from(fileTimestamps.keys()) : Object.keys(fileTimestamps);
+
+        const changedFiles = fileNames.filter((watchfile) => {
+            const prevTimestamp = this.prevTimestamps[watchfile];
+            const nextTimestamp = fileTimestamps.has ? fileTimestamps.get(watchfile) : fileTimestamps[watchfile];
+            this.prevTimestamps[watchfile] = nextTimestamp;
+            return (prevTimestamp || this.startTime) < (nextTimestamp || Infinity)
+        });
 
-        return changedFiles.length === 0 || this.containsOwnDependency(changedFiles);
+        return changedFiles.length !== 0 && this.containsOwnDependency(changedFiles);
```

Comment:
i'm not sure if the logic here is correct. let me know if I misinterpreted the intended behavior of this method.